tm2/pkg/sdk/auth/params.go

```diff
@@ -151,6 +151,68 @@ func (ak AccountKeeper) GetParams(ctx sdk.Context) Params {
 	return params
 }
 
+// WillSetParam defines what needs to be done when the parameter is set.
 func (ak AccountKeeper) WillSetParam(ctx sdk.Context, key string, value any) {
-	// XXX validate input?
+	logger := ak.Logger(ctx)
+	switch key {
+	case "p:unrestricted_addrs":
+		addrs, ok := value.([]string)
+		if !ok {
+			return
+		}
+		ak.applyUnrestrictedAddrsChange(ctx, addrs)
+	default:
+		// No-op for unrecognized keys
+		logger.Error("No-op for unrecognized keys", "key", key)
+	}
+}
+
+func (ak AccountKeeper) applyUnrestrictedAddrsChange(ctx sdk.Context, newAddrs []string) {
+	params, ok := ctx.Value(AuthParamsContextKey{}).(Params)
+	if !ok {
+		panic("missing or invalid AuthParams in context")
+	}
+	// Build sets once.
+	oldSet := make(map[string]struct{}, len(params.UnrestrictedAddrs))
+	for _, addr := range params.UnrestrictedAddrs {
+		oldSet[addr.String()] = struct{}{}
+	}
+	newSet := make(map[string]struct{}, len(newAddrs))
+	for _, s := range newAddrs {
```

Comment:
can we just pass in the deltas(addition/removal), so we don't need to loop and compare against the full old list?